src/app/clusters/ota-requestor/GenericOTARequestorDriver.cpp

```diff
@@ -320,17 +321,48 @@ void GenericOTARequestorDriver::StopDefaultProviderTimer()
         this);
 }
 
-// Returns the next available Provider location
+/**
+ * Returns the next available Provider location. The algorithm is to simply loop through the list of DefaultOtaProviders and return
+ * the next value (based on the last used provider). If no suitable candidate is found, FALSE is returned.
+ */
+
 bool GenericOTARequestorDriver::DetermineProviderLocation(ProviderLocation::Type & providerLocation)
 {
+    // Iterate through the default providers list and find the last used provider. If found, return the provider after it
     auto iterator = mRequestor->GetDefaultOTAProviderListIterator();
-    while (iterator.Next())
+    while (mLastUsedProvider.HasValue() && iterator.Next())
+    {
+        if (ProviderLocationsEqual(iterator.GetValue(), mLastUsedProvider.Value()))
+        {
+            if (iterator.Next())
+            {
+                providerLocation = iterator.GetValue();
+                return true;
+            }
+        }
+    }
+
+    // If no suitable candidate found, return in the following order:
+    // - first element of the default providers list OR
+    // - the last used provider OR
+    // - error (no suitable candidate)
+    iterator = mRequestor->GetDefaultOTAProviderListIterator();
+    if (iterator.Next())
     {
-        // For now, just return the first one
         providerLocation = iterator.GetValue();
         return true;
     }
-
+    else if (mLastUsedProvider.HasValue())
+    {
+        providerLocation = mLastUsedProvider.Value();
+        return true;
+    }
+    else
+    {
+        ChipLogError(SoftwareUpdate,
+                     "GenericOTARequestorDriver::DetermineProviderLocation: No suitable OTA Provider candidate found");
```

Comment:
nit: I think we try to avoid printing the function/file names out, e.g. don't use __FUNC__, etc

Reply:
Right, it unnecessarily increases flash usage, so it's better to keep logs short and concise.

Reply:
Ok, I can change that. Regarding the point about increased RAM usage, we can have compile-time options to change the log-level and thus remove these constant string symbols, no?

Reply:
yes, but this is error-level and even constrained device may take advantage of error logging when debugging things. :)